### ISSUE TITLE
[FEATURE] Grouped Transformations

### ISSUE THREAD

**Issue by david26694** [koaning/scikit-lego#383]
I think it'd be nice to have the following feature, or something similar: The group mean featurizer transforms a numeric feature based on a categorical feature, by deleting the mean of the feature in each category. That way, if we know that the numeric and the categorical feature are very correlated, we can obtain some kind of decorrelated feature.

I think it makes sense to be used both in supervised learning and anomaly detection. I've used it in several supervised problems and heard about it on Kaggle.

I'll give an example. The numeric variable might be the size of an apartment and the category might be the city where the apartment is. These features might be useful to predict apartment prices, for instance. 

The bigger an apartment, the more expensive it should be. However, if we have data from London and Cardiff, London has more expensive and small apartments. The data might actually tell us that smaller apartments are very expensive. If we pass the apartment size and the city through the group mean featurizer, we can mitigate this issue, as the apartment size will be in relation to the mean of the city. 

An apartment of 100 squared meters will be huge in London and regular in Cardiff. For this reason, the feature will give a very high value for a London apartment and a regular value for a Cardiff apartment.

![image](https://user-images.githubusercontent.com/22721584/86916205-8bb2fa00-c123-11ea-9d91-45544094a145.png)

I think it is suited as a transformer, since the idea is to obtain the mean per group at training time in order to avoid leakage.

We might have other use-cases, such:
- Number of downloads of a movie with respect to its genre.
- Amount spent by a customer with respect to some marketing segmentation.

I am aware this is kind of a freaky trick and not many people may want to use it, so let me know if you think it makes sense to include it or not.

**Comment by koaning** [comment]
Interesting, we may already allow for such a "freaky" trick. I also don't think it is *that* freaky; what you

We currently feature a `GroupedEstimator` (docs [here](https://scikit-lego.readthedocs.io/en/latest/meta.html#Grouped-Estimation) which will split the dataset up into groups and then train a model on each (you can do this independantly or with hierarchy in mind). These models however need to be predictors, ie. have a `.predict()` method. There's an example [here](https://scikit-lego.readthedocs.io/en/latest/meta.html#Decayed-Estimation) where it is used to calculate the mean of each month.

Now this grouped estimator will produce a prediction on `y` but you could use this prediction as a feature for another model using our`EstimatorTransformer`(api doc [here](https://scikit-lego.readthedocs.io/en/latest/_modules/sklego/meta/estimator_transformer.html)). Effectively this turns a predictor into a transformer, meaning that the predicted values of an estimator can become features for another model. There's a demo of this [here](https://scikit-lego.readthedocs.io/en/latest/_modules/sklego/meta/estimator_transformer.html). 

If I read what you're suggesting, I think you can construct it via; 

```{python}
from sklearn.dummy import DummyRegressor
from sklego.meta import GroupedEstimator, EstimatorTransformer

grouped_model = GroupedEstimator(DummyRegressor(strategy="mean"), groups=["city"])
group_mean_step = EstimatorTransformer(grouped_model)
```

The `grouped_model` will contain the average per `"city"` this way and the `group_mean_step` will be a transformer as far as scikit-learn is concerned. This means that it can be used as a feature. Potentially via something like; 

```
pipeline = Pipeline([
    ("grab_columns", ColumnSelector(["x1", "x2", "city"])),
    ("ml_features", FeatureUnion([
        ("p1", Pipeline([
            ("grab1", ColumnSelector(["city"])),
            ("mod1", group_mean_step)
        ])),
        ("p2", Pipeline([
            ("grab2", ColumnSelector(["x1", "x2])),
        ]))
    ]))
])

pipeline.fit(df, y).transform(df)
```

Does this suffice?

**Comment by koaning** [comment]
One thing to add here -> you can actually extend this idea not to calculate the mean per group, but to have a regression. If you consider the example for house prices;

![image](https://user-images.githubusercontent.com/1019791/86950806-f0d11480-c150-11ea-9ed0-d164563c35a4.png)


Why keep it linear? 

Another idea for house-prices by the way. You can make the price/m2 regression grouped per city, but perhaps also by house-type. It is plausible that the effect of house size plays out differently for an appartment in an inner city than for a mansion in the countryside.

**Comment by david26694** [comment]
Hmm, thanks for commenting about the `GroupedEstimator` and `EstimatorTransformer`, I was not aware of them and they do look powerful. I think there's something I don't completely understand though.

By looking at the code

```
grouped_model = GroupedEstimator(DummyRegressor(strategy="mean"), groups=["city"])
group_mean_step = EstimatorTransformer(grouped_model)

...

pipeline.fit(df, y).transform(df)

```

what `group_mean_step` acts as is like a target or mean encoder for city (when it is fitted, it computes the mean of the target for each city).

Therefore, if we fit a model with the output of `pipeline` the final model will be something like

```
y = b_0 + b_1 * x_1 + b_2 * x_2 + b_3 * y.groupby(['city']).mean()
```

I was thinking about something like:

```
y = b_0 + b_1 * (x_1 - x_1.groupby(['city']).mean()) + b_2 * (x_2 - x_2.groupby(['city']).mean())
```

perhaps I didn't explain it properly, but I was referring to something which you could define like "normalizing a numeric variable to a categorical one". 

Is this what you understood or am I missing something?

I am trying to think about doing it with the ideas you proposed, the issue that I have is that I would like to fit `group_mean_step` with `x_1 ~ city` instead of `y ~ city`, and I don't know if this is easy to do within a pipeline.

**Comment by koaning** [comment]
Ahh now I see. That you're suggesting there is indeed different than what I mentioned. 

Would I be correct by saying that it sounds like a grouped `StandardScaler` then? In scikit-learn you can specify that the standardscaler only applies the mean and not the standard deviation. I like the idea of having the grouping still be independant of the scaling that you pass in. This should make it much more flexible. 

I think at the moment our `GroupedEstimator` does not allow for transformers like `StandardScaler` because we assume the estimator that goes in only uses `.predict()` instead of `.transform()`. I think we could accomodate for it to be able to though. 

@MBrouns to me this sounds like a valid feature, got anythoughts?

**Comment by david26694** [comment]
Yeah, I couldn't have put it better myself. 

A `GroupedTransformer` would allow for this and more general things. 

For instance, as a similar idea, at some point I tried to implement a  `QuantileGroupTransformer`, which would give the [scikit-learn quantile](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.QuantileTransformer.html#sklearn.preprocessing.QuantileTransformer) by group, and this would be very easy using this `GroupedTransformer`.

**Comment by koaning** [comment]
Now that I think of it, I think we should make a seperate object. The current implementation of the `GroupedEstimator` also allows for hierarchical combinations. It can apply shrinkage to ensure that if a subgroup is small that it can learn more from the parent. This is something that is relatively easy to pull off with predictions, but really unfeasible with transformations. 

This does leave me to wonder what the nice implementation might be. There seem to be a few options. 

1. Keep `GroupedEstimator` as is and introduce a `GroupedTransformer`. 
2. Make the `GroupedEstimator` more general to allow for transformations as well and introduce a `HierarchicalEstimator` to handle the hierarchy. 

@MBrouns I'm leaning towards `GroupedTransformer` but this makes me realise that technically it would be better to have `GroupedEstimator` -> `GroupedPredictor`. Transformers are technically also estimators.

**Comment by MBrouns** [comment]
This looks fun, I think I could have a lot of fun playing around with a GroupedTransformer meta-transformer. 

I agree that we should probably not touch the `GroupedEstimator` at this point, the implementation is quite convoluted to account for the shrinkage and as you said I don't see how that's supposed to work on a transformer yet. I'd suggest building a `GroupedTransformer`, renaming `GroupedEstimator` to `GroupedPredictor` and deprecating the old `GroupedEstimator`

**Comment by pim-hoeven** [comment]
I would like working on the `GroupedTransfomer`, sounds very useful! I'll give it a try today / tomorrow and will keep you posted.

Completely agree that the `GroupedEstimator` became quite a beast with the shrinkage (sorry), so will keep it as-is for now. Maybe I find things to simplify while working on the `GroupedTransformer`, but will make separate issues if I do.
